### ISSUE TITLE
Ring members return as unhealthy and stuck the cluster

### ISSUE THREAD

**Issue by DylanGuedes** [grafana/dskit#207]
**Problem**

On Loki (and apparently in other projects) we're facing this weird scenario where ring members return as unhealthy and once this happens to more than one member, the ring gets stuck (maybe because we're running with replication_factor = 3? :thinking: )

What we know so far is:
- Clicking manually to forget the unhealthy members does fix the issue
- I don't think we ever faced this with typical healthy rollouts; I think all occasions it did happen were when pods were deleted or OOMKilled or the k8s node hosting the pods had an outage, so maybe it has a relationship with our heartbeat logic :thinking: 
- For Loki, this happens for both, ingesters members and distributors members, so it isn't related to a specific implementation of the project
- The typical log line we use to identify this behavior is the one below (for Loki):
```
level=warn ts=2022-07-28T17:32:28.308736477Z caller=grpc_logging.go:43 method=/httpgrpc.HTTP/Handle duration=179.426µs err="rpc error: code = Code(500) desc = at least 2 live replicas required, could only find 1 - unhealthy instances: x.x.x.x:9095,y.y.y.y:9095\n" msg=gRPC
```

This is probably not enough detail to track down what is wrong, so if this happens again I'll make sure to grab a memory dump and other things that might help.

**Comment by pracucci** [comment]
> the ring gets stuck

What does it mean? Can you be more specific what you mean by "stuck"?

**Comment by DylanGuedes** [comment]
> > the ring gets stuck
> 
> What does it mean? Can you be more specific what you mean by "stuck"?

sorry, this was a few months back so I don't remember all the details. but by "stuck" I meant that all ring operations would start to error reporting that it had too many unhealthy instances.

**Comment by DylanGuedes** [comment]
trying to illustrate better what I had in mind when I created this:
- let's say you have 15 members in the ring, all of them working healthy and just fine. we ran into a scenario where, because 2 members were unhealthy, all other members start to error reporting that the ring had too many unhealthy instances.
I can't remember though if this was a misconfiguration on our side. but in our case, we'd be better if the other 13 healthy instances kept working normally
